server/server.go

```diff
@@ -60,19 +121,121 @@ func (d *DataAvailabilityLightClient) SubmitBlock(ctx context.Context, blockReq
 		return &dalc.SubmitBlockResponse{
 			Result: &dalc.DAResponse{
 				Code:    dalc.StatusCode_STATUS_CODE_ERROR,
-				Message: fmt.Sprintf("failed to submit tx: code %d", resp.Code),
+				Message: fmt.Sprintf("failed to submit tx: code %d: %s", resp.Code, resp.RawLog),
 			},
 		}, err
 	}
 
+	err = d.heightMapper.Save(int64(blockReq.Block.Header.Height), resp.Height)
+	if err != nil {
+		// TODO: we proabably need to do something more drastic than warn if a
+		// block we just submitted already exists in the height mapper
+		d.logger.Error(err.Error())
+		return nil, err
+	}
+
 	d.logger.Info("Submitted block to celstia", "height", resp.Height, "gas used", resp.GasUsed, "hash", resp.TxHash)
 	return &dalc.SubmitBlockResponse{Result: &dalc.DAResponse{Code: dalc.StatusCode_STATUS_CODE_SUCCESS}}, nil
 }
 
-func (d *DataAvailabilityLightClient) CheckBlockAvailability(context.Context, *dalc.CheckBlockAvailabilityRequest) (*dalc.CheckBlockAvailabilityResponse, error) {
-	return nil, nil
+// CheckBlockAvailability samples shares from the underlying data availability layer
+func (d *DataAvailabilityLightClient) CheckBlockAvailability(ctx context.Context, req *dalc.CheckBlockAvailabilityRequest) (*dalc.CheckBlockAvailabilityResponse, error) {
+	// check if the block we're trying to submit already has a celestia height associated with it
+	optimintHeight := int64(req.Header.Height)
+	celestiaHeight, has := d.heightMapper.Search(optimintHeight)
+	if !has {
+		return nil, NoAssociatedBlockError{optimintHeight}
+	}
+
+	// get the dah for the block
+	dah, err := getDAH(ctx, d.node.CoreClient, celestiaHeight)
+	if err != nil {
+		return nil, err
+	}
+
+	err = d.node.ShareServ.SharesAvailable(ctx, dah)
+	switch err {
+	case nil:
+		return &dalc.CheckBlockAvailabilityResponse{
+			Result: &dalc.DAResponse{
+				Code: dalc.StatusCode_STATUS_CODE_SUCCESS,
+			},
+			DataAvailable: true,
+		}, nil
+	default:
+		return &dalc.CheckBlockAvailabilityResponse{
+			Result: &dalc.DAResponse{
+				Code:    dalc.StatusCode_STATUS_CODE_UNSPECIFIED,
+				Message: err.Error(),
+			},
+			DataAvailable: false,
+		}, err
+	}
+}
+
+func (d *DataAvailabilityLightClient) RetrieveBlock(ctx context.Context, req *dalc.RetrieveBlockRequest) (*dalc.RetrieveBlockResponse, error) {
+	// check if the block we're trying to submit already has a celestia height associated with it
+	optimintHeight := int64(req.Height)
+	if _, has := d.heightMapper.Search(optimintHeight); !has {
+		return nil, NoAssociatedBlockError{optimintHeight}
+	}
+
+	dah, err := getDAH(ctx, d.node.CoreClient, int64(req.Height))
+	if err != nil {
+		return nil, err
+	}
```

Comment:
same comment as above, this will have to be changed when we refactor to actually use celestia-node as intented

Reply:
https://github.com/celestiaorg/dalc/issues/40